stub/src/main/java/io/grpc/stub/ClientResponseObserver.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.stub;
+
+import io.grpc.ExperimentalApi;
+
+/**
+ * Specialization of {@link StreamObserver} implemented by clients in order to interact with the
+ * advanced features of a call such as flow-control.
+ */
+@ExperimentalApi
+public interface ClientResponseObserver<ReqT, RespT> extends StreamObserver<RespT> {
+  /**
+   * Called by the runtime at the start to provide a reference to the {@link CallStreamObserver}
+   * for the outbound stream. This can  be used to listen to onReady events,
+   * disable auto inbound flow and perform other advanced functions.
+   *
+   * <p>{@link CallStreamObserver#setOnReadyHandler(Runnable)} and
+   * {@link CallStreamObserver#disableAutoInboundFlowControl()} may only be called within this
+   * callback.
+   *
+   * <pre>
+   *   someStub.fullDuplexCall(new ClientResponseObserver() {
+   *     public void onStart(final CallStreamObserver&lt;ReqT, RespT&gt; requestStream) {
+   *       requestStream.setOnReadyHandler(new Runnable() {
+   *          public void run() {
+   *            while(requestStream.isReady() && someIterator.hasNext()) {
+   *              requestStream.onNext(someIterator.next());
+   *            }
+   *       });
+   *   });
+   * </pre>
+   */
+  void onStart(final CallStreamObserver<ReqT> requestStream);
```

Comment:
We might need to tweak the name `onStart` or put in the documentation that you can _only_ call `setOnReadyHandler` and `disableAutoInboundFlowControl` within `onStart` since start() hasn't actually been called yet (can't send a message, for instance). It's sort of `beforeStart`.